table.go

```diff
@@ -230,7 +287,21 @@ func (rt *RoutingTable) removePeer(p peer.ID) {
 	if bucket.remove(p) {
 		// peer removed callback
 		rt.PeerRemoved(p)
-		return
+
+		// remove this bucket if it was the last bucket and it's now empty
+		// provided it isn't the ONLY bucket we have.
+		if len(rt.buckets) > 1 && bucketID == len(rt.buckets)-1 && len(bucket.peers()) == 0 {
+			rt.buckets[bucketID] = nil
```

Comment:
Might be just blanking here, but why are we doing this? Does it help the garbage collector out in some way?

Reply:
It helps the GC garbage collect the struct pointed to by the slice element as I understand from here:

https://github.com/golang/go/wiki/SliceTricks (Look at the section on `Delete`).
However, to truly grok the mechanics and reasons behind this, I've posted a question on one of our go forums and will post the answer here when I get one.